apps/frontend/src/app/PageTeam/CharacterDisplay/Tabs/TabOptimize/index.tsx

```diff
@@ -1024,6 +967,107 @@ export default function TabBuild() {
   )
 }
 
+function LevelFilter({
+  levelTotal,
+  levelLow,
+  levelHigh,
+  generatingBuilds,
+  optConfigId,
+}: {
+  levelTotal: string
+  levelLow: number
+  levelHigh: number
+  generatingBuilds: boolean
+  optConfigId: string
+}) {
+  const database = useDatabase()
+  const { t } = useTranslation('page_character_optimize')
+  return useMemo(
+    () => (
+      <CardLight>
+        <CardContent sx={{ display: 'flex', gap: 1 }}>
+          <Typography sx={{ fontWeight: 'bold' }}>{t`levelFilter`}</Typography>
+          <SqBadge color="info">{levelTotal}</SqBadge>
+        </CardContent>
+        <Divider />
+        <CardContent>
+          <ArtifactLevelSlider
+            levelLow={levelLow}
+            levelHigh={levelHigh}
+            setLow={(levelLow) =>
+              database.optConfigs.set(optConfigId, { levelLow })
+            }
+            setHigh={(levelHigh) =>
+              database.optConfigs.set(optConfigId, { levelHigh })
+            }
+            setBoth={(levelLow, levelHigh) =>
+              database.optConfigs.set(optConfigId, {
+                levelLow,
+                levelHigh,
+              })
+            }
+            disabled={generatingBuilds}
+          />
+        </CardContent>
+      </CardLight>
+    ),
+    [
+      database,
+      generatingBuilds,
+      levelHigh,
+      levelLow,
+      levelTotal,
+      optConfigId,
+      t,
+    ]
+  )
+}
+
+function PanelLeft({ characterKey }: { characterKey: CharacterKey }) {
```

Comment:
I would recommend using `React.memo` https://react.dev/reference/react/memo, also, probably rename this to `CharacterPanel` or something.

Reply:
I'm unsure about this since the function name disappears from stack traces